godot-core/crates/godot-core-macros/src/lib.rs

```diff
@@ -0,0 +1,360 @@
+                let return_value_metadata = Sig::param_metadata(-1);
+
+                // Arguments meta-information
+                let argument_count = NUM_ARGS as u32;
+                let mut arguments_info: [sys::GDExtensionPropertyInfo; NUM_ARGS] = godot::private::gdext_get_arguments_info!(Sig, #( #param_idents, )*);
+                let mut arguments_metadata: [sys::GDExtensionClassMethodArgumentMetadata; NUM_ARGS]
+                    = std::array::from_fn(|i| Sig::param_metadata(i as i32));
+
+                let class_name = StringName::from(stringify!(#class_name));
+                let method_name = StringName::from(stringify!(#method_name));
+
+                let method_info = sys::GDExtensionClassMethodInfo {
+                    name: method_name.string_sys(),
+                    method_userdata: std::ptr::null_mut(),
+                    call_func: Some(varcall_func),
+                    ptrcall_func: Some(ptrcall_func),
+                    method_flags: #method_flags as u32,
+                    has_return_value: has_return_value as u8,
+                    return_value_info: std::ptr::addr_of_mut!(return_value_info_sys),
+                    return_value_metadata,
+                    argument_count,
+                    arguments_info: arguments_info.as_mut_ptr(),
+                    arguments_metadata: arguments_metadata.as_mut_ptr(),
+                    default_argument_count: 0,
+                    default_arguments: std::ptr::null_mut(),
+                };
+
+                godot::private::out!("   Register fn:   {}::{}", stringify!(#class_name), stringify!(#method_name));
+                sys::interface_fn!(classdb_register_extension_class_method)(
+                    sys::get_library(),
+                    class_name.string_sys(),
+                    std::ptr::addr_of!(method_info),
+                );
+            }
+        }.into()
+}
+
+/// Returns a C function which acts as the callback when a virtual method of this instance is invoked.
+//
+// There are currently no virtual static methods. Additionally, virtual static methods dont really make a lot
+// of sense. Therefore there is no need to support them.
+
+#[proc_macro]
+pub fn gdext_virtual_method_callback(input: TokenStream) -> TokenStream {
+    let method_def = parse_macro_input!(input as MethodDef);
+    let method_signature = method_def.sig;
+
+    let instance_ptr = quote! {instance_ptr};
+    let args = quote! {args};
+    let ret = quote! {ret};
+
+    let ptrcall = get_ptrcall(
+        &method_def.class_name,
+        method_signature,
+        instance_ptr,
+        args,
+        ret,
+    );
+
+    quote! { Some({
+        use godot::sys;
+
+        unsafe extern "C" fn function(
+            instance_ptr: sys::GDExtensionClassInstancePtr,
+            args: *const sys::GDExtensionConstTypePtr,
+            ret: sys::GDExtensionTypePtr,
+        ) {
+            #ptrcall
+        }
+        function
+    })
+    }
```

Comment:
rustfmt ignores code inside macro bodies, but could you still make sure they are properly formatted?
One way to achieve this is to write `{ }` instead of `quote! { }` and then run rustfmt...

This also applies to some `quote!` statements below.